logger/logger.go

```diff
@@ -114,52 +129,31 @@ func (l *logger) LogMode(level LogLevel) Interface {
 
 // Info print info
 func (l logger) Info(ctx context.Context, msg string, data ...interface{}) {
-	if l.LogLevel >= Info {
-		l.Printf(l.infoStr+msg, append([]interface{}{utils.FileWithLineNum()}, data...)...)
-	}
+	l.printf(ctx, l.infoStr, msg, Info, data...)
 }
 
 // Warn print warn messages
 func (l logger) Warn(ctx context.Context, msg string, data ...interface{}) {
-	if l.LogLevel >= Warn {
-		l.Printf(l.warnStr+msg, append([]interface{}{utils.FileWithLineNum()}, data...)...)
-	}
+	l.printf(ctx, l.warnStr, msg, Warn, data...)
 }
 
 // Error print error messages
 func (l logger) Error(ctx context.Context, msg string, data ...interface{}) {
-	if l.LogLevel >= Error {
-		l.Printf(l.errStr+msg, append([]interface{}{utils.FileWithLineNum()}, data...)...)
-	}
+	l.printf(ctx, l.errStr, msg, Error, data...)
 }
 
 // Trace print sql message
 func (l logger) Trace(ctx context.Context, begin time.Time, fc func() (string, int64), err error) {
 	if l.LogLevel > Silent {
 		elapsed := time.Since(begin)
 		switch {
+		case l.LogLevel == Info:
```

Comment:
This case should be the last case, and maybe the method maybe better refactor like:

```go
switch {
case: ...
  ft = xxx;
  trace = xxx;
default:
  return
}

sql, rows := fc()
if rows == -1 {
   l.Printf(ft, utils.FileWithLineNum(), trace, float64(used.Nanoseconds())/1e6, "-", sql)
} else {
  l.Printf(ft, utils.FileWithLineNum(), trace, float64(used.Nanoseconds())/1e6, rows, sql)
}
```

Reply:
I have not considered it, and the number of formatting parameters of Trace has not been considered.